Direct link to media files are not handled properly (anymore?)

With file:
https://download.kiwix.org/zim/wikipedia/wikipedia_en_geography_novid_2019-04.zim

Open article "Canton of Zürich", click on blue link "Zürich" on the first line. The file targeted is an audio file, but is not in the ZIM... so why the link exists?

Zim Check is showing this:
```
[ERROR] Invalid internal links found :
  upload.wikimedia.org/wikipedia/commons/8/8e/Al_arabic.ogg (//upload.wikimedia.org/wikipedia/commons/8/8e/Al_arabic.ogg) was not found in article A/Arabic
  upload.wikimedia.org/wikipedia/commons/0/02/En-us-Archaea.ogg (//upload.wikimedia.org/wikipedia/commons/0/02/En-us-Archaea.ogg) was not found in article A/Archaea
  upload.wikimedia.org/wikipedia/commons/0/0e/En-uk-Birmingham.ogg (//upload.wikimedia.org/wikipedia/commons/0/0e/En-uk-Birmingham.ogg) was not found in article A/Birmingham
  upload.wikimedia.org/wikipedia/commons/b/b7/Pt-br-República_Federativa_do_Brasil.ogg (//upload.wikimedia.org/wikipedia/commons/b/b7/Pt-br-Rep%C3%BAblica_Federativa_do_Brasil.ogg) was not found in article A/Brazil
  upload.wikimedia.org/wikipedia/commons/6/68/Sv-Carl_von_Linné.ogg (//upload.wikimedia.org/wikipedia/commons/6/68/Sv-Carl_von_Linn%C3%A9.ogg) was not found in article A/Carl_Linnaeus
  upload.wikimedia.org/wikipedia/commons/d/dc/En-us-fungi.ogg (//upload.wikimedia.org/wikipedia/commons/d/dc/En-us-fungi.ogg) was not found in article A/Fungus
  upload.wikimedia.org/wikipedia/commons/5/58/De-Bundesrepublik_Deutschland.ogg (//upload.wikimedia.org/wikipedia/commons/5/58/De-Bundesrepublik_Deutschland.ogg) was not found in article A/Germany
  upload.wikimedia.org/wikipedia/commons/2/26/En-ca-Hong_Kong.oga (//upload.wikimedia.org/wikipedia/commons/2/26/En-ca-Hong_Kong.oga) was not found in article A/Hong_Kong

```
So looks like the files aren't referenced (and maybe not downloaded properly)

--- a/src/util/rewriteUrls.ts
+++ b/src/util/rewriteUrls.ts
@@ -1,6 +1,6 @@
 import * as urlParser from 'url';
 import { isMirrored } from './saveArticles';
-import { migrateChildren, getMediaBase, getFullUrl } from './misc';
+import { migrateChildren, getMediaBase, getFullUrl, getRelativeFilePath } from './misc';
 import { redirectsXId } from '../stores';
 import { Dump } from '../Dump';
 import MediaWiki from '../MediaWiki';
@@ -50,6 +50,10 @@ export async function rewriteUrl(articleId: string, mw: MediaWiki, dump: Dump, l
         href = `${wikiProtocol}${href}`;
         linkNode.setAttribute('href', href);
         hrefProtocol = urlParser.parse(href).protocol;
+    }
+
+    if (!rel && linkNode.getAttribute('resource')) {
+        rel = 'mw:MediaLink';
     }
 
     if (!href) {
@@ -115,9 +119,13 @@ export async function rewriteUrl(articleId: string, mw: MediaWiki, dump: Dump, l
                 lat = parts[4];
                 lon = parts[5];
             } else if (rel === 'mw:MediaLink') {
-                if (!dump.nopdf && /\.pdf/i.test(href)) {
+                const shouldScrape = href.includes('.pdf') && !dump.nopdf ||
+                    href.includes('.ogg') && !dump.nopic && !dump.novid && !dump.nodet;
+
+                if (shouldScrape) {
                     try {
-                        linkNode.setAttribute('href', getMediaBase(href, true));
+                        const newHref = getRelativeFilePath(articleId, getMediaBase(href, true), 'I');
+                        linkNode.setAttribute('href', newHref);
                         mediaDependencies.push(href);
                     } catch (err) {
                         logger.warn('Error parsing url:', err);

--- a/src/util/saveArticles.ts
+++ b/src/util/saveArticles.ts
@@ -244,6 +244,7 @@ function widthXHeightSorter(a: DominoElement, b: DominoElement) {
 }
 
 async function treatVideo(mw: MediaWiki, dump: Dump, srcCache: KVS<boolean>, articleId: string, videoEl: DominoElement): Promise<{ mediaDependencies: string[] }> {
+    // This function handles audio tags as well as video tags
     const webUrlHost = urlParser.parse(mw.webUrl).host;
     const mediaDependencies: string[] = [];
     // Worth noting:
@@ -251,9 +252,6 @@ async function treatVideo(mw: MediaWiki, dump: Dump, srcCache: KVS<boolean>, art
     // When it's only audio, there will be a single OGG file
     // For video, we get multiple SOURCE tages with different resolutions
 
-    const posterUrl = videoEl.getAttribute('poster');
-    const videoPosterUrl = getFullUrl(webUrlHost, posterUrl, mw.base);
-    const newVideoPosterUrl = getRelativeFilePath(articleId, getMediaBase(videoPosterUrl, true), 'I');
     let videoSources: any[] = Array.from(videoEl.children).filter((child: any) => child.tagName === 'SOURCE');
 
     // Firefox is not able to display correctly <video> nodes with a height < 40.
@@ -270,14 +268,6 @@ async function treatVideo(mw: MediaWiki, dump: Dump, srcCache: KVS<boolean>, art
         return { mediaDependencies };
     }
 
-    if (posterUrl) { videoEl.setAttribute('poster', newVideoPosterUrl); }
-    videoEl.removeAttribute('resource');
-
-    if (!srcCache.hasOwnProperty(videoPosterUrl)) {
-        srcCache[videoPosterUrl] = true;
-        mediaDependencies.push(videoPosterUrl);
-    }
-
     videoSources = videoSources.sort(widthXHeightSorter);
 
     const sourcesToRemove = videoSources.slice(1); // All but first
@@ -292,6 +282,22 @@ async function treatVideo(mw: MediaWiki, dump: Dump, srcCache: KVS<boolean>, art
     if (!fileBase) {
         DU.deleteNode(sourceEl);
         return;
+    }
+
+    /* Remove useless 'resource' attribute */
+    videoEl.removeAttribute('resource');
+
+    const posterUrl = videoEl.getAttribute('poster');
+    if (posterUrl) {
+        const videoPosterUrl = getFullUrl(webUrlHost, posterUrl, mw.base);
+        const newVideoPosterUrl = getRelativeFilePath(articleId, getMediaBase(videoPosterUrl, true), 'I');
+        if (posterUrl) { videoEl.setAttribute('poster', newVideoPosterUrl); }
+        videoEl.removeAttribute('resource');
+
+        if (!srcCache.hasOwnProperty(videoPosterUrl)) {
+            srcCache[videoPosterUrl] = true;
+            mediaDependencies.push(videoPosterUrl);
+        }
     }
 
     const newUrl = getRelativeFilePath(articleId, fileBase, 'I');
@@ -438,10 +444,10 @@ export async function treatMedias(parsoidDoc: DominoElement, mw: MediaWiki, dump
     let mediaDependencies: string[] = [];
     /* Clean/rewrite image tags */
     const imgs = Array.from(parsoidDoc.getElementsByTagName('img'));
-    const videos: DominoElement = Array.from(parsoidDoc.getElementsByTagName('video'));
+    const videos: DominoElement = Array.from(parsoidDoc.querySelectorAll('video, audio'));
     const srcCache: KVS<boolean> = {};
 
-    for (const videoEl of videos) {
+    for (const videoEl of videos) { // <video /> and <audio />
         const ret = await treatVideo(mw, dump, srcCache, articleId, videoEl);
         mediaDependencies = mediaDependencies.concat(ret.mediaDependencies);
     }

--- a/test/unit/urlRewriting.test.ts
+++ b/test/unit/urlRewriting.test.ts
@@ -30,10 +30,11 @@ test('Url re-writing', async (t) => {
     const $nonScrapedWikiLink = makeLink($doc, '/wiki/this_page_does_not_exist', '', 'fake link');
     const $specialMap = makeLink($doc, '/wiki/Special:Map/9/51.51/-0.08/en', '', 'Interactive map outlining London');
     const $hashLink = makeLink($doc, '#cite_note-LAS-150', '', 'The London Air Ambulance');
+    const $resourceLink = makeLink($doc, '//upload.wikimedia.org/wikipedia/commons/c/c6/De-Z%C3%BCrich.ogg', '', 'De-Z%C3%BCrich.ogg', 'Zurich', { resource: './Media:De-Zürich.ogg' });
 
     await rewriteUrl(parentArticleId, mw, dump, $geo);
     t.assert($geo.nodeName === 'A', 'Geo is still a link');
-    t.equal($geo.getAttribute('href'), 'geo:37.786971,-122.399677', 'Geo HREF is correct'); // seems to be a bug in Domino which forces ://
+    t.equal($geo.getAttribute('href'), 'geo:37.786971,-122.399677', 'Geo HREF is correct');
 
     await rewriteUrl(parentArticleId, mw, dump, $geoHack);
     t.assert($geoHack.nodeName === 'A', 'GeoHack is still a link');
@@ -77,5 +78,9 @@ test('Url re-writing', async (t) => {
 
     await rewriteUrl(parentArticleId, mw, dump, $nonScrapedWikiLink);
     t.assert(!$nonScrapedWikiLink.parentElement, 'nonScrapedWikiLink has been deleted');
+
+    await rewriteUrl(parentArticleId, mw, dump, $resourceLink);
+    t.assert($resourceLink.nodeName === 'A', 'resourceLink is still a link');
+    t.equal($resourceLink.getAttribute('href'), '../I/m/De-Z%C3%BCrich.ogg', 'resourceLink has been re-written');
 
 });

--- a/test/util.ts
+++ b/test/util.ts
@@ -7,12 +7,16 @@ export function leftPad(_num: number, length: number) {
     return '0'.repeat(length - num.length) + num;
 }
 
-export function makeLink($doc: Document, href: string, rel: string, title: string, text: string = href) {
+export function makeLink($doc: Document, href: string, rel: string, title: string, text: string = href, attributes: KVS<string> = {}) {
     const $link = $doc.createElement('a');
     $link.setAttribute('href', href);
     $link.setAttribute('rel', rel);
     $link.setAttribute('title', title);
     $link.innerHTML = text;
+
+    for (const [key, value] of Object.entries(attributes)) {
+        $link.setAttribute(key, value);
+    }
 
     const $wrapper = $doc.createElement('div');
     $wrapper.appendChild($link);

